Source/WebKit/WebProcess/GPU/media/AudioVideoRendererRemote.cpp

```diff
@@ -0,0 +1,839 @@
+        RefPtr gpuProcessConnection = m_gpuProcessConnection.get();
+        if (!isGPURunning() || !gpuProcessConnection)
+            return MediaTimePromise::createAndReject(PlatformMediaError::Cancelled);
+
+        m_state.currentTime = time;
+        return gpuProcessConnection->connection().sendWithPromisedReply<MediaPromiseConverter>(Messages::RemoteAudioVideoRendererProxyManager::SeekTo(m_identifier, time), 0);
+    });
+}
+
+bool AudioVideoRendererRemote::seeking() const
+{
+    return m_state.seeking;
+}
+
+void AudioVideoRendererRemote::setPreferences(VideoMediaSampleRendererPreferences preferences)
+{
+    RefPtr gpuProcessConnection = m_gpuProcessConnection.get();
+    if (!isGPURunning() || !gpuProcessConnection)
+        return;
+
+    gpuProcessConnection->connection().send(Messages::RemoteAudioVideoRendererProxyManager::SetPreferences(m_identifier, preferences), 0);
+}
+
+void AudioVideoRendererRemote::setHasProtectedVideoContent(bool isProtected)
+{
+    RefPtr gpuProcessConnection = m_gpuProcessConnection.get();
+    if (!isGPURunning() || !gpuProcessConnection)
+        return;
+
+    gpuProcessConnection->connection().send(Messages::RemoteAudioVideoRendererProxyManager::SetHasProtectedVideoContent(m_identifier, isProtected), 0);
+}
+
+AudioVideoRendererRemote::TrackIdentifier AudioVideoRendererRemote::addTrack(TrackType type)
+{
+    // the sendSync() call requires us to run on the connection's dispatcher, which is the main thread.
+    assertIsMainThread();
+    // FIXME: Uses a new Connection for remote playback, and not the main GPUProcessConnection's one.
+    // FIXME: m_mimeTypeCache is a main-thread only object.
+    auto sendResult = m_gpuProcessConnection.get()->connection().sendSync(Messages::RemoteAudioVideoRendererProxyManager::AddTrack(m_identifier, type), 0);
+    auto result = std::get<0>(sendResult.takeReplyOr(makeUnexpected(PlatformMediaError::IPCError)));
+    ASSERT(!!result);
+    return *result;
+}
+
+void AudioVideoRendererRemote::removeTrack(TrackIdentifier trackIdentifier)
+{
+    RefPtr gpuProcessConnection = m_gpuProcessConnection.get();
+    if (!isGPURunning() || !gpuProcessConnection)
+        return;
+
+    gpuProcessConnection->connection().send(Messages::RemoteAudioVideoRendererProxyManager::RemoveTrack(m_identifier, trackIdentifier), 0);
+}
+
+void AudioVideoRendererRemote::enqueueSample(TrackIdentifier trackIdentifier, Ref<MediaSample>&& sample, std::optional<MediaTime> expectedMinimum)
+{
+    RefPtr gpuProcessConnection = m_gpuProcessConnection.get();
+    if (!isGPURunning() || !gpuProcessConnection)
+        return;
+    gpuProcessConnection->connection().send(Messages::RemoteAudioVideoRendererProxyManager::EnqueueSample(m_identifier, trackIdentifier, MediaSamplesBlock::fromMediaSample(sample), expectedMinimum), 0);
+    if (auto it = m_requestMediaDataWhenReadyData.find(trackIdentifier); it != m_requestMediaDataWhenReadyData.end())
+        it->value.pendingSamples++;
```

Comment:
Is there a possibility that we enqueue before filling `m_requestMediaDataWhenReadyData`? What is the impact?

Reply:
calling any of those callback with a nullptr is equivalent of cancelling the callback.